### ISSUE TITLE
Don't swallow session commit exceptions or make it configurable

### ISSUE THREAD

**Issue by Tratcher** [aspnet/Docs#3380]
_From @cwe1ss on May 18, 2017 11:37_

When the session middleware fails to persist a session (e.g. because the database is not available), it logs the exception and [swallows it](https://github.com/aspnet/Session/blob/dev/src/Microsoft.AspNetCore.Session/SessionMiddleware.cs#L120). 

The request will then continue normally, which obviously leads to very unpredictable behavior.

A typical example:
* Someone stores a shopping basket in session. The user adds an item but the commit fails. The app doesn't know about this so it will happily report a "The item has been added" message, which obviously isn't true.

I don't think the session middleware should have the authority do decide that sessions are "non-critical" and optional. This decision can only be made by an application developer.

I think you should either completely remove this error handling or at least make it configurable (with *not catching exceptions* as the default).

_Copied from original issue: aspnet/Session#165_

**Comment by Tratcher** [comment]
By the time the middleware calls CommitAsync the response has already been sent and it's too late report an error to the client.

The recommended way to check for such errors is to call `await feature.Session.CommitAsync();` from app code when you're done writing to the session. Then you can do what you like with the error. The same goes for calling LoadAsync.

**Comment by Tratcher** [comment]
_From @cwe1ss on May 18, 2017 11:58_

Thx for the quick response! If I call CommitAsync() on my own, will the session middleware call it again or is it smart enough to not commit again?

I think it would be great if this behavior was documented on the [app-state docs page](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/app-state) - it's a pretty important thing to know.

**Comment by Tratcher** [comment]
Yes, it's smart enough not to call commit twice (if it succeeds).
https://github.com/aspnet/Session/blob/c3ca9cf3342eb66d966ee681c193bfd22b6b701a/src/Microsoft.AspNetCore.Session/DistributedSession.cs#L217

I'll move this to a doc issue.

**Comment by Rick-Anderson** [comment]
Add info in here to [Introduction to session and application state in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/app-state)

**Comment by Rick-Anderson** [comment]
@cwe1ss  can you draft some text we can add to [Introduction to session and application state in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/app-state) ?
